### DISCUSSION TITLE
Feature Request: Browser-Based LLM inference

### DISCUSSION THREAD

**Discussion by gloryknight** [open-webui/open-webui#8149] [Ideas]
# Feature Request

I propose integrating https://github.com/ngxson/wllama, a powerful open-source library for WebAssembly binding for llama.cpp, enabling in-browser LLM inference as a selectable model option within the web UI.  This integration would allow users to run GGUF models, directly within their browser.

Local execution could be implemented as a checkbox in the model capabilities (see admin panes->models->click on model).

By leveraging wllama, we can combine the user-friendly interface of open-webui with local inference in the browser. This offers several potential benefits for our users:

* **Hosting on low end servers:** Processing LLM queries locally within the browser could allow hosting of open-webui without any connections to external LLM providers.
* **Mature interface for local inference:** Mature interface of open-webui will enhance experience of users interested in using in-browser inference.
* **Offline access:** Users could continue to utilize LLM capabilities even without an active internet connection.

wllama is available under the permissive MIT License, making it a suitable choice for integration into our existing web application.

**Comment by aviallon** [comment]
I massively support this idea! WebAssembly models would also be awesome to use for conversation naming/auto-tagging, for instance.
It could massively reduce load (and latency) on the back-end servers, and it would **automatically scale** with the number of clients!

**Comment by AABoyles** [comment]
[Implemented a version of this](https://github.com/AABoyles/open-webui/tree/feat/browser-runtime) using the [WebLLM API](https://webllm.mlc.ai/docs/), since it's a very close mirror to the OpenAI API. Is anyone still interested? Should I submit a PR to dev?
